### PR TITLE
Add dafny; fixup Boogie

### PR DESCRIPTION
**Pull request by layus**
#### Motivation for the change

Introduce dafny, which uses Boogie.
And fix Boogie, because the last edit (changing the baseName) right before merge (#22400) broke it.
Also propote boogie and dafny as toplevel packages, because they both have toplevel standalone executables corresponding to their names.

#### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/cc @obadz ('coz it's dotnet), @taktoa ('coz I changed Boogie), @pSub ('coz he approved the changes to Boogie) and @rycee ('coz he merged Boogie)

### CODE REVIEW

File: pkgs/top-level/dotnet-packages.nix [hunk 1]
```diff
@@ -221,8 +221,9 @@ let self = dotnetPackages // overrides; dotnetPackages = with self; {
   # SOURCE PACKAGES
 
   Boogie = buildDotnetPackage rec {
-    baseName = "Boogie-unstable";
+    baseName = "Boogie";
     version = "2017-01-03";
+    name = "${baseName}-unstable-${version}";
```
Comment (rycee) [inline pkgs/top-level/dotnet-packages.nix]: Looks good to me 👍

File: pkgs/top-level/dotnet-packages.nix [hunk 2]
```diff
@@ -251,6 +252,50 @@ let self = dotnetPackages // overrides; dotnetPackages = with self; {
...
+    postFixup = ''
+      for lib in ${Boogie}/lib/dotnet/${Boogie.baseName}/*.dll{,.mdb}; do
+        ln -s $lib $out/lib/dotnet/${baseName}/
+      done
+      # We generate our own executable scripts
+      rm -f $out/lib/dotnet/${baseName}/dafny{,-server}
+    '';
+
+    meta = {
+      description = "A programming language with built-in specification constructs.";
```
Comment (pSub) [inline pkgs/top-level/dotnet-packages.nix]: Descriptions should not end with a period.

File: pkgs/top-level/dotnet-packages.nix [hunk 3]
```diff
@@ -251,6 +252,50 @@ let self = dotnetPackages // overrides; dotnetPackages = with self; {
...
+      homepage = "http://research.microsoft.com/dafny";
```
Comment (pSub) [inline pkgs/top-level/dotnet-packages.nix]: Nitpick: You don't need quotes around Urls.

File: pkgs/top-level/dotnet-packages.nix (hunk 2 above)
Comment (layus) [inline pkgs/top-level/dotnet-packages.nix]: Right! Fixed.

File: pkgs/top-level/dotnet-packages.nix (hunk 3 above)
Comment (layus) [inline pkgs/top-level/dotnet-packages.nix]: Unless it is a merge-blocking issue for you, I am going to leave this as-is. There is nothing saying that urls should not be quoted in the CONTRIBUTING.md. And nothing in the nix manual about urls as language values, only paths.

**Comment by layus** [comment]
Fixed according to reviews, and removed the usage of 'lib' that made travis fail.

**Pull request merged**